controllers/machinepool_controller.go

```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(mp, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always reconcile the Status.Phase field.
+		r.reconcilePhase(mp)
+
+		// Always attempt to Patch the MachinePool object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, mp); err != nil {
+			if reterr == nil {
+				reterr = err
+			}
+		}
+	}()
+
+	// Reconcile and retrieve the Cluster object.
+	if mp.Labels == nil {
+		mp.Labels = make(map[string]string)
+	}
+	mp.Labels[clusterv1.ClusterLabelName] = mp.Spec.ClusterName
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, mp.ObjectMeta.Namespace, mp.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get cluster %q for machine %q in namespace %q",
+			mp.Labels[clusterv1.ClusterLabelName], mp.Name, mp.Namespace)
+	}
+
+	// Handle deletion reconciliation loop.
+	if !mp.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, cluster, mp)
+	}
+
+	// Handle normal reconciliation loop.
+	return r.reconcile(ctx, cluster, mp)
+}
+
+func (r *MachinePoolReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster, mp *clusterv1.MachinePool) (ctrl.Result, error) {
+	// If the MachinePool belongs to a cluster, add an owner reference.
+	if cluster != nil && r.shouldAdopt(mp) {
+		mp.OwnerReferences = util.EnsureOwnerRef(mp.OwnerReferences, metav1.OwnerReference{
+			APIVersion: cluster.APIVersion,
+			Kind:       cluster.Kind,
+			Name:       cluster.Name,
+			UID:        cluster.UID,
+		})
+	}
+
+	// If the MachinePool doesn't have a finalizer, add one.
+	if !util.Contains(mp.Finalizers, clusterv1.MachinePoolFinalizer) {
+		mp.ObjectMeta.Finalizers = append(mp.ObjectMeta.Finalizers, clusterv1.MachinePoolFinalizer)
+	}
+
+	// Call the inner reconciliation methods.
+	reconciliationErrors := []error{
+		r.reconcileBootstrap(ctx, mp),
+		r.reconcileInfrastructure(ctx, mp),
+		r.reconcileNodeRefs(ctx, cluster, mp),
+	}
+
+	// Parse the errors, making sure we record if there is a RequeueAfterError.
+	res := ctrl.Result{}
+	errs := []error{}
+	for _, err := range reconciliationErrors {
+		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
+			// Only record and log the first RequeueAfterError.
+			if !res.Requeue {
+				res.Requeue = true
+				res.RequeueAfter = requeueErr.GetRequeueAfter()
+				klog.Infof("Reconciliation for MachinePool %q in namespace %q asked to requeue: %v", mp.Name, mp.Namespace, err)
+			}
+			continue
+		}
+
+		errs = append(errs, err)
+	}
+	return res, kerrors.NewAggregate(errs)
+}
+
+func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, mp *clusterv1.MachinePool) (ctrl.Result, error) {
+	if mp.Spec.ProviderIDs == nil {
+		return ctrl.Result{}, errNilProviderID
```

Comment:
Do we actually need to guard against a nil list here? What if the list never gets initialized, and we issue a delete? Wouldn't the delete get stuck because of the finalizer?